R/collect.R

```diff
@@ -60,9 +84,217 @@ collect_predictions <- function(x) {
     rlang::abort(msg)
   }
 
-  collector(x, coll_col = coll_col)
+  x <- filter_predictions(x, parameters)
+
+  if (summarize) {
+    x <- average_predictions(x, parameters)
+  } else {
+    x <- collector(x, coll_col = coll_col)
+  }
+  x
+}
+
+filter_predictions <- function(x, parameters) {
+  if (is.null(parameters)) {
+    return(x)
+  }
+  metrics <- attr(x, "metrics")
+  params <- attr(x, "parameters")
+  if (is.null(metrics) | is.null(params)) {
+    rlang::warn(
+      paste(
+        strwrap(
+          paste("The object is missing some attributes; it is probably from",
+                "an earlier version of `tune`. The predictions can't be filtered." ),
+          prefix = ""
+        ),
+        collapse = "\n"
+      )
+    )
+
+    return(x)
+  }
+
+  param_names <- params$id
+  parameters <- dplyr::select(parameters, dplyr::one_of(param_names))
+  if (ncol(parameters) != length(param_names)) {
+    rlang::abort(
+      paste0("`parameters` should only have columns: ",
+             paste0("'", param_names, "'", collapse = ", "))
+    )
+  }
+  x$.predictions <-
+    purrr::map(x$.predictions, dplyr::inner_join, parameters, by = param_names)
+  x
+}
+
+# Functions for summarizing the various results over replicate out-of-sample
+#  predictions. For regression cases, the numeric predictions are simply
+#  averaged. For classification models, the problem is more complex. When class
+#  probabilities are used, these are averaged and then re-normalized to make
+#  sure that they add to one. If hard class predictions also exist in the data,
+#  then these are determined from the summarized probability estimates (so that
+#  they match). If only hard class predictions are in the results, then the
+#  mode is used to summarize.
+
+numeric_summarize <- function(x, p, wider = TRUE) {
+  pred_cols <- grep("^\\.pred", names(x), value = TRUE)
+  nms <- names(x)
+  group_cols <- nms[!(nms %in% pred_cols)]
+  x <-
+    x %>%
+    tidyr::pivot_longer(
+      cols = c(dplyr::one_of(pred_cols)),
+      names_to = ".column",
+      values_to = ".value"
+    ) %>%
+    dplyr::group_by(!!!rlang::syms(group_cols), .column) %>%
+    dplyr::summarize(.value = mean(.value, na.rm = TRUE))
```

Comment:
Should you `ungroup()` here? Right now you just `ungroup()` if `wider = TRUE`. If `wider = FALSE` this will likely return a grouped_df. Not sure if you want that

Reply:
Hmm maybe you do want it for `prob_and_class_summarize()` later on?

Reply:
Yes. Otherwise I group again there.